### PR TITLE
[Zoe] Add multipool autoswap

### PR DESCRIPTION
**Pull request by katelynsills**
To add a new token (and therefore a new liquidity pool with the new token and the central token), the user calls the public `addPool` method on the `publicAPI` with the new token issuer and a keyword. 

```js
    const addMoolaPoolResult = await E(publicAPI).addPool(
      moolaR.issuer,
      'Moola',
    );

    const addSimoleansPoolResult = await E(publicAPI).addPool(
      simoleanR.issuer,
      'Simoleans',
    );
```

The user can see the current liquidity pool allocations (how much liquidity is available) by querying by the secondary token brand:

```js
    t.deepEquals(
      await E(publicAPI).getPoolAllocation(moolaR.brand),
      harden({
        Moola: moolaR.amountMath.getEmpty(),
        CentralToken: centralR.amountMath.getEmpty(),
        MoolaLiquidity: moolaLiquidityAmountMath.getEmpty(),
      }),
      `The poolAmounts object should only have keywords for Moola, CentralToken, and MoolaLiquidity. Values should be empty`,
    );
```

The user can get the current price:

```js
    // Bob looks up the price of 3 moola in central tokens
    const priceInCentralTokens = bobAutoswap.getPrice(moola(3), centralR.brand);
    t.deepEquals(priceInCentralTokens, centralTokens(1));
```

The user can swap tokens:

```js
    const bobMoolaForCentralProposal = harden({
      want: { CentralToken: centralTokens(1) },
      give: { Moola: moola(3) },
    });
    const bobMoolaForCentralPayments = harden({ Moola: bobMoolaPayment });

    const { seat: bobSeat, payout: bobPayoutP } = await zoe.redeem(
      bobExclInvite,
      bobMoolaForCentralProposal,
      bobMoolaForCentralPayments,
    );

    // Bob swaps
    const offerOk = bobSeat.swap();
    t.equal(offerOk, 'Swap successfully completed.');
```

The user can also check the price between two secondary tokens and swap between secondary tokens in the same way as trades with the central token:

```js
    const priceInMoola = await E(bobPublicAPI).getPrice(
      simoleans(74),
      moolaR.brand,
    );
    const bobSimsForMoolaProposal = harden({
      want: { Moola: moola(10) },
      give: { Simoleans: simoleans(74) },
    });

    ... 
    bobSeatSimsForMoola.swap();
```

The two swaps that occur under the hood when trading one secondary token for another, occur synchronously, meaning that no slippage can occur while the swap is happening.

Users (or wallets) can learn what keywords to use by calling `getBrandKeywordRecord` or `getKeywordForBrand`:

```js
    t.deepEquals(
      await E(bobPublicAPI).getBrandKeywordRecord(),
      harden({
        Moola: moolaR.brand,
        Simoleans: simoleanR.brand,
        CentralToken: centralR.brand,
        MoolaLiquidity: await E(moolaLiquidityIssuer).getBrand(),
        SimoleansLiquidity: await E(simoleanLiquidityIssuer).getBrand(),
      }),
      `keywords have expected brands`,
    );

    t.equals(
      await E(bobPublicAPI).getKeywordForBrand(moolaR.brand),
      'Moola',
      `moola keyword is Moola`,
    );
```

TODO:

- [x]  Swapping between two secondary tokens should be synchronous
- [x]  Show how a wallet or users generally can get the mapping from brands to keywords to learn what the keywords are

### CODE REVIEW

**Comment by dtribble** [comment]
Looks very nice, and a fine API for discussion/demonstration. 

You've probably already considered this: I suspect that we could run into API usability in actual usage by other contracts because the names of keys are dynamic. That seems a fine trade for the clarity now, and we can look further either the "seat params" spike

**Comment by katelynsills** [comment]
> You've probably already considered this: I suspect that we could run into API usability in actual usage by other contracts because the names of keys are dynamic. That seems a fine trade for the clarity now, and we can look further either the "seat params" spike

I'll try to provide an example of this in the test code, but I think it's actually not bad. The contract can have a `publicAPI` method mapping `keywords` to `brands` and/or `issuers`, which can be used as a map to look up the keyword given an ERTP brand.

**Comment by katelynsills** [comment]
removing @erights and adding @DavidBruant as reviewer

File: packages/zoe/src/contracts/multipoolAutoswap.js
Comment (Chris-Hibbert) [inline packages/zoe/src/contracts/multipoolAutoswap.js]: should this be `getCurrentPrice()`? It's just an instantaneous quote, with no hint that the price will remain stable.

File: packages/zoe/test/unitTests/contractSupport/test-bondingCurves.js
Comment (Chris-Hibbert) [inline packages/zoe/test/unitTests/contractSupport/test-bondingCurves.js]: Please add something about what might be wrong if this test fails.

**Review by Chris-Hibbert — CHANGES_REQUESTED** [review]
The code looks straightforward, which is reassuring. There are some large functions which would be more readable if some internal code could be extracted to functions.

I'd really like to see a test of the multipool, showing both that the payouts are reasonable and that the internal pools are well-behaved. I suspect that doing multi-step trades, and checking the prices afterwards would suffice.

**Comment by katelynsills** [comment]
> I'd really like to see a test of the multipool, showing both that the payouts are reasonable and that the internal pools are well-behaved. I suspect that doing multi-step trades, and checking the prices afterwards would suffice.

@Chris-Hibbert Hmm, can you explain a little more of what you are looking for? There is a clear test of the multipool, when Bob gives simoleans and gets moola (line 426 of test-multipoolAutoswap.js). I also check the pool allocations afterward. Was there something else you were looking for?

**Comment by Chris-Hibbert** [comment]
> can you explain a little more of what you are looking for? 

I see it now. GitHub collapsed it because it was too big. When is this ever the right default? Can I change a setting somewhere so it defaults to showing me files with lots of new code?  `<scowl>`

File: packages/zoe/src/contracts/multipoolAutoswap.js
Comment (katelynsills) [reply packages/zoe/src/contracts/multipoolAutoswap.js]: Hmm, this is from an old version. Resolving

File: packages/zoe/src/contracts/multipoolAutoswap.js
Comment (katelynsills) [reply packages/zoe/src/contracts/multipoolAutoswap.js]: Sure, I can move this logic into a function, but I think there are two goals here: to verify that a central token brand exists, and then assign it to the variable `centralTokenBrand` for the rest of the contract to use.

File: packages/zoe/src/contracts/multipoolAutoswap.js
Comment (katelynsills) [reply packages/zoe/src/contracts/multipoolAutoswap.js:103]: Yup, good idea

File: packages/zoe/src/contracts/multipoolAutoswap.js
Comment (katelynsills) [reply packages/zoe/src/contracts/multipoolAutoswap.js]: No reason, I'll change it to `do`

Comment (katelynsills) [reply packages/zoe/src/contracts/multipoolAutoswap.js]: No, under the hood, `getPrice` always favors the pool, so we are already accounting for rounding.

**Comment by Chris-Hibbert** [comment]
This is a good test of the happy path. I'd like to see some tests of the edge cases. Maybe not all of these, and perhaps you can come up with other interesting cases for a TODO list.

 * a trader asks for less than they can get (i.e. a limit order in case the price changes)
 * trader asks for more than they can get (i.e. the price moved against them)?
 * a liquidity pool is drained, then someone tries to trade it
 * error cases:
    * requesting trades for tokens that are not present
    * getPrice(centralToken, centralToken)
    * attempt to trade with liquidity token

File: packages/zoe/src/contracts/multipoolAutoswap.js
Comment (katelynsills) [reply packages/zoe/src/contracts/multipoolAutoswap.js]: And whether it ends up as two or three cases, this is certainly a situation where I would `// eslint-disable` and use
> 
> ```
>  if () {
>  } else if () {
>  } else if () {
>  }
> ```

Can you explain this more? Since the code `returns`, `else` seems unnecessary

File: packages/zoe/src/contracts/multipoolAutoswap.js
Comment (katelynsills) [reply packages/zoe/src/contracts/multipoolAutoswap.js]: Sure, that sounds good. I'll change autoswap too.

File: packages/zoe/src/contracts/multipoolAutoswap.js
Comment (katelynsills) [reply packages/zoe/src/contracts/multipoolAutoswap.js]: Good point, I'll leave both out

File: packages/zoe/src/contracts/multipoolAutoswap.js
Comment (Chris-Hibbert) [reply packages/zoe/src/contracts/multipoolAutoswap.js]: Even with the returns, I find 
```
 if (<condition A>) {
    .....
   return;
 } else if (<condition B>) {
    .....
   return;
 } else if (<condition C>) {
    .....
   return;
 }
```
far more readable than

```
 if (<condition A>) {
    .....
   return;
 }
 if (<condition B>) {
    .....
   return;
 }
 if (<condition C>) {
    .....
   return;
 }
```

Particularly when the bodies are long or there are more than a couple of them. If they're really exclusive, and nothing intervenes between the various bodies, then I can evaluate them separately.  If there are a series of separate conditionals, I have to look if any code has been added between conditions A and B in order to understand the context for condition C. This is also why I turn such things into switch/case when the conditions are all testing the same value--as a reader, I can make simpler assumptions about the context for each branch.

It seems like the only reason to prefer making them separate is that it makes it possible to insert code between the cases.

**Comment by katelynsills** [comment]
> This is a good test of the happy path. I'd like to see some tests of the edge cases. Maybe not all of these, and perhaps you can come up with other interesting cases for a TODO list.
> 
> * a trader asks for less than they can get (i.e. a limit order in case the price changes)
> * trader asks for more than they can get (i.e. the price moved against them)?
> * a liquidity pool is drained, then someone tries to trade it
> * error cases:
>   
>   * requesting trades for tokens that are not present
>   * getPrice(centralToken, centralToken)
>   * attempt to trade with liquidity token

I like these suggestions, but the happy path is all we are testing right now for Zoe contracts. Would it be ok if we make an issue of out this comment to tackle this in a separate PR?

**Comment by Chris-Hibbert** [comment]
If the policy is not to write test code for error cases and edge conditions then there should certainly be an issue to address that.

Is that also true of Zoe itself?

File: packages/zoe/src/contracts/multipoolAutoswap.js
Comment (katelynsills) [reply packages/zoe/src/contracts/multipoolAutoswap.js]: Sure, I just divided this up into separate functions.

**Comment by katelynsills** [comment]
> If the policy is not to write test code for error cases and edge conditions then there should certainly be an issue to address that.

> Is that also true of Zoe itself?

There's no policy stating that, of course. I'm trying to point out that it doesn't make sense today to handle all the edge cases for multipoolAutoswap, when we both know that other contracts have only one happy path test. Rather, I think it makes more sense to get this PR up to the same level as other contracts, and since we have many tasks still to go for the hackathon release, file an issue and come back to the edge cases. 

Re: Zoe, as we both know, Zoe is pre-alpha, so we certainly do not claim to have covered all, or even most edge cases in tests yet.

**Review by Chris-Hibbert — APPROVED** [review]
Thanks! This looks great.

**Comment by katelynsills** [comment]
Waiting for #853 to land before landing this

**Pull request merged**
